language/move-lang/tests/move_check/typing/bind_pop_resource.exp

```diff
@@ -12,20 +12,6 @@ error:
    │     -------- Is found to be a non-copyable type here
    │
 
-error: 
-
-   ┌── tests/move_check/typing/bind_pop_resource.move:6:13 ───
-   │
- 6 │         let _r: R = R{};
-   │             ^^ Cannot ignore resource values. The value must be used
-   ·
- 6 │         let _r: R = R{};
-   │                 - The type: '0x8675309::M::R'
-   ·
- 2 │     resource struct R {}
-   │     -------- Is found to be a non-copyable type here
-   │
-
```

Comment:
This should still be an error. what's going on here?

Reply:
Or at least there should be a new error that says "unused resource in local _r"

Reply:
You are right. This should still be an error.

If I just put
```
module M {
    resource struct R {}

    fun t0() {
        let _r: R = R{};
    }
}
```

It gives
```
error:

   ┌── test.move:5:24 ───
   │
 5 │         let _r: R = R{};
   │                        ^ Invalid return
   ·
 5 │         let _r: R = R{};
   │             -- The local '_r' still contains a resource value due to this assignment. The resource must be consumed before the function returns
   │

```

But if I put `let _: R = R{};` above or below `let _r`, the error is gone...

Please allow me a bit more time to check this.

Reply:
I think it's because I just rely on unused assignments to check that. You might need another case for:
- unused _ variables that are resourceful

Reply:
I guess things might be a bit more complex there, even without the underscore, the `r` is not warned.
```
module M {
    resource struct R {}

    fun t0() {
        let _: R = R{};
        let r: R = R{};
    }
}
```

Reply:
That's very bizarre... If you need help debugging this let me know

Reply:
My suspicion is that the pass `LocalsSafety` is shortcut by unused assignment but let me look more into this. I think I can debug this one.

Reply:
```
        PassResult::Naming(nprog, errors) => {
            let (tprog, errors) = typing::translate::program(nprog, errors);
            check_errors(errors)?;
            ^^^^^^^^^^^^^^^^^^^^^^
            run(PassResult::Typing(tprog), until)
        }
        PassResult::Typing(tprog) => {
            let (hprog, errors) = hlir::translate::program(tprog);
            run(PassResult::HLIR(hprog, errors), until)
        }
        PassResult::HLIR(hprog, errors) => {
            let (cprog, errors) = cfgir::translate::program(errors, hprog);
            check_errors(errors)?;
            run(PassResult::CFGIR(cprog), until)
        }
        PassResult::CFGIR(cprog) => {
            let compiled_units = to_bytecode::translate::program(cprog)?;
            assert!(until == Pass::Compilation);
            run(PassResult::Compilation(compiled_units), Pass::Compilation)
        }
```

This `check_error()` short-circuited the execution such that the CFGIR (where `LocalSafety` lives) is not executed.

Removing it brings back the error, but I doubt that this is the right way to fix it
```
error:

   ┌── test.move:5:13 ───
   │
 5 │         let r: R = R{};
   │             ^ Unused assignment or binding for local 'r'. Consider removing it or replacing/prefixing it with '_'
   │

error:

   ┌── test.move:6:16 ───
   │
 6 │     let _: R = R{};
   │                   ^ Invalid return
   ·
 5 │         let r: R = R{};
   │             - The local 'r' still contains a resource value due to this assignment. The resource must be consumed before the function returns
   │

```

Reply:
This might not be easy to fix, as reasoning whether a local is actually used or not needs CFGIR, but reasoning whether a `_` binds to a resource just needs `typing`.

This issue no impact on the overall correctness of the compiler. But it will leave certain errors (e.g., those checks that requires HLIR, CFGIR) undetected if there are naming or typing isssues.

Reply:
Oh if its just an issue with that test its fine. It is well known that you do not get unused resource warnings until after typing has completed. This is normal and expected.

But you will still need to fix that one case for resources in that one test as I mentioned above

Reply:
I think this is also the reason why that error is no longer showing in the `bind_pop_resource.move` case (i.e., the case you mentioned above)
```
$cat bind_pop_resource.move
module M {
    resource struct R {}

    fun t0() {
        let _: R = R{};
        let _r: R = R{};
        let (_, _):(R, R) = (R{}, R{});
    }
}
```

After making `_r` as binding instead of ignore, that error will require CFGIR to check while the others did not pass the `typing` pass. Hence the original error is not printed.

I think it's better for me to put a comment in `bind_pop_resource.move` to mark that this `let _r` is also wrong, but its error won't be shown. Or better still, create another test in `locals` with only `let _r: R = R{};` in `fun t0()`.

Reply:
I would do both of those things, put a comment and make another test. Thanks!